File: wraplayout/src/main/java/autoexclue/PageManager.java
package autoexclue;

import android.util.SparseArray;

import java.util.ArrayList;
import java.util.List;


public class PageManager {
     AutoExcludeLayout layout;
     List<CementItem<?>> data;
     List<Segment> segments;
     SparseArray<Integer> findPageMap;
     Segment curSegment;
     AutoExcludeLayout.Adapter adapter = new AutoExcludeLayout.Adapter();
     int pageCount = 0;

    public void setAdapter(AutoExcludeLayout.Adapter adapter) {
        this.adapter = adapter;
    }

    private void setData(List<CementItem<?>> data) {
        this.data = data;
    }


    private void switchToItem(CementItem<?> item) {
        int find = data.indexOf(item);
        if (find == -1) return;
        if (find >= 0 && find < data.size()) {
            switchToPage(find);
        }
    }

    private void switchToPage(int dataIndex) {
        int page = findPageByDataIndex(dataIndex);
        curSegment = segments.get(page);
        adapter.clearData();
        List<CementItem<?>> items = new ArrayList<>();
        for (int i = curSegment.start; i < curSegment.end; i++) {
            items.add(data.get(i));
        }
        adapter.addDataList(items);
    }

    private int findPageByDataIndex(int dataIndex) {
        int page = 0;
        Segment cur = segments.get(page);
        while (page < segments.size() && cur != null && cur.end < dataIndex) {
            page++;
            cur = segments.get(page);
        }
        return page;
    }

}

File: wraplayout/src/main/java/autoexclue/PageManager.kt
package autoexclue

import android.util.SparseArray
import java.util.*

class PageManager {
    var layout: AutoPagerView? = null
    var data: List<AbstractCellItem<*>>? = null
    var segments: List<Segment>? = null
    var findPageMap: SparseArray<Int>? = null
    var curSegment: Segment? = null
    var adapter = AutoPageAdapter()
    var pageCount = 0
    private fun switchToItem(item: AbstractCellItem<*>) {
        val find = data!!.indexOf(item)
        if (find == -1) return
        if (find >= 0 && find < data!!.size) {
            switchToPage(find)
        }
    }

    private fun switchToPage(dataIndex: Int) {
        val page = findPageByDataIndex(dataIndex)
        curSegment = segments!![page]
        adapter.clearData()
        val items: MutableList<AbstractCellItem<*>> = ArrayList()
        for (i in curSegment!!.start until curSegment!!.end) {
            items.add(data!![i])
        }
        adapter.addDataList(items)
        adapter.notifyDataSetChanged()
    }

    private fun findPageByDataIndex(dataIndex: Int): Int {
        var page = 0
        var cur = segments!![page]
        while (page < segments!!.size && cur != null && cur.end < dataIndex) {
            page++
            cur = segments!![page]
        }
        return page
    }
}